rosbag2_cpp/include/rosbag2_cpp/message_definitions/message_definition_cache.hpp

```diff
@@ -72,16 +49,39 @@ class MessageDefinitionCache final
 public:
   /**
    * Concatenate the message definition with its dependencies into a self-contained schema.
-   * The format is different for MSG and IDL definitions, and is described fully here:
-   * [MSG](https://mcap.dev/specification/appendix.html#ros2msg-data-format)
-   * [IDL](https://mcap.dev/specification/appendix.html#ros2idl-data-format)
+   * The format is different for MSG and IDL definitions, and is described fully in
+   * docs/message_definition_encoding.md
    * Throws DefinitionNotFoundError if one or more definition files are missing for the given
    * package resource name.
    */
-  ROSBAG2_STORAGE_MCAP_PUBLIC
-  std::pair<Format, std::string> get_full_text(const std::string & package_resource_name);
+  rosbag2_storage::MessageDefinition get_full_text(const std::string & package_resource_name);
+
+  enum struct Format
+  {
+    MSG,
+    IDL,
+  };
 
 private:
+  struct MessageSpec
+  {
+    MessageSpec(Format format, std::string text, const std::string & package_context);
+    const std::set<std::string> dependencies;
+    const std::string text;
+    Format format;
+  };
+
+  struct DefinitionIdentifier
+  {
+    Format format;
+    std::string package_resource_name;
+
+    bool operator==(const DefinitionIdentifier & di) const
+    {
+      return (format == di.format) && (package_resource_name == di.package_resource_name);
+    }
```

Comment:
I think it would be more appropriate and efficient if  
```
    Format format;
    std::string package_resource_name;
```
will be a private members and we will have getters for them.
The rational for that is that their are settling up only once on object construction and using then as read only.
First of all it is not safe to have direct access to member variables which are supposed to be read only.
We can occasionally overwrite them.
Another point it would be more efficient if we will ad one more optimization and cache hash calculation. i.e. Calculate hash once on constructing and store it in private member variable. In this case
`struct DefinitionIdentifierHash` will be simplified to 
```
  struct DefinitionIdentifierHash
  {
    std::size_t operator()(const DefinitionIdentifier & di) const
    {
       return di.hash()
    }
  }
```
It will be more efficient operations in and unorderd_maps and sets